Search preview version 2024-05-01-preview

sazeesha098 opened the pull request:
# Data Plane API - Pull Request
<!-- 🚨🚨🚨 Replace this line with a summary and reason for these changes to your data plane API 🚨🚨🚨 -->

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).

* Link to API Spec engagement record issue:

Is this review for (select one):

- [ ] a private preview

### Change Scope

<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous API Spec document (if applicable), and the root paths that have been updated. </sup>
* Design Document:
* Previous API Spec Doc:
* Updated paths:

### Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

### Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/pr-suppressions) 
to get approval.

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling

 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://github.com/Azure/azure-api-style-guide/blob/main/README.md)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications

 * [Azure REST API Guidelines](https://aka.ms/azapi/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/azapi/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links

 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

<details>
  <summary>Checks stuck in `queued` state?</summary>
If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.  
</details>

fix https://github.com/Azure/azure-rest-api-specs/issues/27237

sazeesha committed:
Base commit

sazeesha committed:
Update api version to 2024-05-01-Preview

sazeesha committed:
supporting image vectorization with BYOE

sazeesha committed:
AML vectorizer definition

sazeesha committed:
Add vector weights

sazeesha committed:
Add hybridSearch query time parameter

sazeesha committed:
Add new results combination profile we are adding to the scoring profile

sazeesha committed:
Add vector query thresholding

sazeesha committed:
Add return subscores for hybrid / vector queries

sazeesha committed:
Add a OneLake data source type

sazeesha committed:
Addressing check failures 1

heaths commented on specification/search/data-plane/Azure.Search/preview/2024-05-01-Preview/searchindex.json:
```diff
@@ -1320,7 +1326,14 @@
         "semantic": {
           "$ref": "#/definitions/SemanticDebugInfo",
           "readOnly": true,
+          "x-nullable": true,
```
Why is this marked nullable? That means the user can pass `null` as a value and is probably not what you mean. This is only useful for json+mergepatch.

heaths commented on specification/search/data-plane/Azure.Search/preview/2024-05-01-Preview/searchindex.json:
```diff
@@ -1423,6 +1436,84 @@
+            "$ref": "#/definitions/TextResult"
+          },
+          "readOnly": true,
+          "x-nullable": true,
```
More here and below.

heaths commented on specification/search/data-plane/Azure.Search/preview/2024-05-01-Preview/searchindex.json:
```diff
@@ -1423,6 +1436,84 @@
+    },
+    "TextResult": {
+      "properties": {
+        "searchScore": {
```
Why isn't this `confidenceScore`?

heaths commented on specification/search/data-plane/Azure.Search/preview/2024-05-01-Preview/searchindex.json:
```diff
@@ -1423,6 +1436,84 @@
+          "type": "number",
+          "format": "double",
+          "readOnly": true,
+          "description": "The vector similarity score for this document. Note this is the canonical definition of similarity metric, not the 'distance' version. For example, cosine similarity instead of cosine distance."
```
What's the expected range? Should be documented. Same for other fields.

robertklee replied:
Different vector similarity metrics have different ranges, so the only invalid values should be Inf, -Inf, and NaN.

For example, dot product and euclidean distance technically have unbounded distance, and dot product can go negative, since neither have a normalization component in it.

robertklee replied:
A \dot B = \sum a_i * b_i and if a_i and b_i are not fixed range, this can be any number.

heaths commented on specification/search/data-plane/Azure.Search/preview/2024-05-01-Preview/searchindex.json:
```diff
@@ -1922,11 +2022,45 @@
+      ],
+      "properties": {
+        "url": {
+          "type": "string",
```
`"format": "url"` (or maybe "uri" - I forget which; the compiler will check)

heaths commented on specification/search/data-plane/Azure.Search/preview/2024-05-01-Preview/searchindex.json:
```diff
@@ -1941,6 +2075,95 @@
+        "value": {
+          "type": "number",
+          "format": "double",
+          "description": "The threshold will filter based on the '@search.score' value. Note this is the @search.score returned as part of the search response. The threshold direction will be chosen for higher @search.score."
```
Range? Document all ranges if not otherwise obvious.

robertklee replied:
Answered in different comment thread :)

BevLoh commented on specification/search/data-plane/Azure.Search/preview/2024-05-01-Preview/searchindex.json:
```diff
@@ -0,0 +1,3095 @@
+        }
+      ],
+      "properties": {
+        "base64EncodedBinary": {
```
API feedback: rename to `base64Image`

mikekistler commented on specification/search/data-plane/Azure.Search/preview/2024-05-01-Preview/searchservice.json:
```diff
@@ -0,0 +1,12693 @@
+            "description": "Set to delimitedText when blobs are plain CSV files."
+          },
+          {
+            "value": "
```
Should be pascal case.

mikekistler replied:
Should be Pascal case.

mikekistler reviewed (changes requested):
searchservice.json is not valid OpenAPI. I flagged the key problems I could find. Please fix and then tag me to re-review.

sazeesha committed:
Apply feedback

mattgotteiner commented on specification/search/data-plane/Azure.Search/preview/2024-05-01-Preview/searchservice.json:
```diff
@@ -0,0 +1,12697 @@
+      },
+      "description": "Specifies the properties for connecting to a user-defined vectorizer."
+    },
+    "AIServicesVisionVectorizer": {
```
Ted - Should this be called vision if it's multimodal?

mattgotteiner replied:
Ted - consider avoiding branding in swagger

mattgotteiner commented on specification/search/data-plane/Azure.Search/preview/2024-05-01-Preview/searchindex.json:
```diff
@@ -0,0 +1,3097 @@
+          "format": "int32",
+          "description": "Determines the maximum number of documents to be retrieved by the text query portion of a hybrid search request. The values needs to be between 1 and 100 000. Default is 1000. A higher value allows paging through more documents, at the cost of a higher latency."
+        },
+        "countAndFacetMode": {
```
Ted / Mike - Document which one is the default. Avoid picking the option "countRetrievableResutls" as this is a breaking change to the current default behavior

mattgotteiner commented on specification/search/data-plane/Azure.Search/preview/2024-05-01-Preview/searchindex.json:
```diff
@@ -0,0 +1,3097 @@
+        "maxTextRecallSize": {
+          "type": "integer",
+          "format": "int32",
+          "description": "Determines the maximum number of documents to be retrieved by the text query portion of a hybrid search request. The values needs to be between 1 and 100 000. Default is 1000. A higher value allows paging through more documents, at the cost of a higher latency."
```
Mike - explain more clearly how this is different than top

mikekistler replied:
My recollection from the API review: this value determines the number of documents to factor into the document weighting, which can affect the document scores. It has nothing to do with how many documents are returned.

mattgotteiner replied:
Thank you! Was just going off the transcript

shmed replied:
I'll try to reformulate this. It does indeed affect how many documents are returned, since it defines the maximum amount of documents that we will consider for the text query. Example, if you set the Value of MaxTextRecallSize to 100, and the K value for the vector query to 10, then the maximum amount of documents that can be returned is 110 (100 + 10, assuming there's no overlap between the text results and the vector results).

mikekistler commented on specification/search/data-plane/Azure.Search/preview/2024-05-01-Preview/searchindex.json:
```diff
@@ -0,0 +1,3097 @@
+    },
+    "QueryResultDocumentVectorSubscores": {
+      "type": "object",
+      "additionalProperties": {
+        "type": "object",
+        "items": {
+          "$ref": "#/definitions/SingleVectorFieldResult"
+        },
+        "readOnly": true
+      },
```
A schema with `type: object` should not specify `items` -- `items` only applies for `type: array`.

shmed replied:
it should actually be an array (since we are going to have one entry per vector-field pair)

mikekistler commented on specification/search/data-plane/Azure.Search/preview/2024-05-01-Preview/searchindex.json:
```diff
@@ -0,0 +1,3097 @@
+        }
+      },
+      "readOnly": true,
+      "additionalProperties": true,
```
You should avoid having additionalProperties in the same schema as named properties. Adding new properties then becomes a breaking change.

mikekistler commented on specification/search/data-plane/Azure.Search/preview/2024-05-01-Preview/searchservice.json:
```diff
@@ -0,0 +1,12697 @@
+            "description": "Set to delimitedText when blobs are plain CSV files."
+          },
+          {
+            "value": "
```
Don't specify additionalProperties and named properties in the same schema.

mikekistler replied:
Don't specify additionalProperties and named properties in the same schema.

sazeesha committed:
Addressing feedback for final review

sazeesha committed:
remove sub scores from swagger

sazeesha committed:
fix Swagger Avocado failed

sazeesha committed:
fix file path

sazeesha committed:
fix ref kind for VectorThreshold

sazeesha committed:
Fix path string

sazeesha committed:
Make modelVersion required for vision vectorization and Add experimental modelName option

sazeesha committed:
Attempting to fix INCONSISTENT_API_VERSION

sazeesha098 commented:
@mikekistler we have addressed all the feedback, and we should be ready to get this merged, can you please do the final review and tag it to be merged?

mikekistler commented on specification/search/data-plane/Azure.Search/preview/2024-05-01-Preview/searchindex.json:
```diff
@@ -0,0 +1,3095 @@
+  "info": {
+    "title": "SearchIndexClient",
+    "description": "Client that can be used to query an index and upload, merge, or delete documents.",
+    "version": "2024-05-01-Preview",
```
I think this change is still necessary.

sazeesha098 replied:
@mikekistler I had made that change and had to revert it as we were getting Swagger Avocado failures "The API version of the swagger is inconsistent with its file path.".

I wanted to keep the folder structure consistent with the previous API versions.

sazeesha098 replied:
I have addressed that now.

mikekistler commented on specification/search/data-plane/Azure.Search/preview/2024-05-01-Preview/searchservice.json:
```diff
@@ -0,0 +1,12432 @@
+  "info": {
+    "title": "SearchServiceClient",
+    "description": "Client that can be used to manage and query indexes and documents, as well as manage other resources, on a search service.",
+    "version": "2024-05-01-Preview",
```
And please also change this one:
```suggestion
    "version": "2024-05-01-preview",
```

sazeesha098 replied:
addressed it

mikekistler reviewed (changes requested):
It is a minor thing, but I would really like you to use lowercase "preview" in your API version.

sazeesha committed:
Undo capitalized "Preview" in the API version

sazeesha committed:
remove the verison temporarily

sazeesha committed:
Adding the preview back

mikekistler reviewed (approved):
Thank you for the fix! 👍

sazeesha098 commented:
Hold merge

sazeesha098 commented:
@JeffreyRichter Do we need the tag for "Approved-BreakingChange" since the CI fails only for cross version.

sazeesha098 committed:
Merge branch 'main' into sazeesha/search-2024-05-01-preview

JeffreyRichter commented:
Why all the breaks shown here: https://github.com/Azure/azure-rest-api-specs/pull/27662/checks?check_run_id=23175608186

sazeesha098 commented:
> Why all the breaks shown here: https://github.com/Azure/azure-rest-api-specs/pull/27662/checks?check_run_id=23175608186
@JeffreyRichter they are comparing this preview to the last stable version.

sazeesha committed:
fix parameter ordering in Documents_SearchGet compared to previous GA

sazeesha098 commented:
> I understand that but why is the service team introducing these breaks? We don't want to break customers. So breaks must all be justified before they can be approved.

I have summarized the changes here:
Errors::

1025 - RequiredStatusChange :
    The 'required' status changed from the old version('False') to the new version('True').
    New: Azure.Search/preview/2024-05-01-preview/searchservice.json#L12241:7
    Old: Azure.Search/stable/2023-11-01/searchservice.json#L10638:7
==> (Will fix in next GA) Added PBI for us to add skillsetCount as required in the next GA version

1033 - RemovedProperty
	The new version is missing a property found in the old version. Was 'code' renamed or removed?
    New: common-types/data-plane/v1/types.json#L52:7
    Old: Azure.Search/stable/2023-11-01/searchindex.json#L1850:7
6 out of 6 occurrence : 
==> (Will fix in next GA) Added PBI to use error definition from common type like in preview version 

1034 - AddedRequiredProperty
	The new version has new required property 'aliasesCount, skillsetCount' that was not found in the old version.
    New: Azure.Search/preview/2024-05-01-preview/searchservice.json#L12225:9
    Old: Azure.Search/stable/2023-11-01/searchservice.json#L10622:9
3 out of 3 occurrences :
==> (Ignore) aliasesCount is only supported in the preview version 
==> (Will fix in next GA) Added PBI for us to add skillsetCount as required in the next GA version

1042 - ChangedParameterOrder
    The order of parameter '$select' was changed.
    New: Azure.Search/preview/2024-05-01-preview/searchindex.json#L87:9
    Old: Azure.Search/stable/2023-11-01/searchindex.json#L90:9
==> (Need Approval) Alligned the ordering in 2024-05-01-preview Vs 2023-11-01, but it will fail with previous preview version, that will need an approval this time (next preview it should not alert)

Warnings::

1017 - ReferenceRedirection
    The '$ref' property points to different models in the old and new versions.
    New: Azure.Search/preview/2024-05-01-preview/searchindex.json#L1107:13
    Old: Azure.Search/stable/2023-11-01/searchindex.json#L1005:13
40 occurrences
==> (Will fix in next GA) Added PBI to use error definition from common type like in preview version 

1020 - AddedEnumValue
    The new version is adding enum value(s) 'Edm.Half, Edm.Int16, Edm.SByte' from the old version.
    New: Azure.Search/preview/2024-05-01-preview/searchservice.json#L3137:5
    Old: Azure.Search/stable/2023-11-01/searchservice.json#L2676:5
11 occurrences
==> (Ignore) New enum supported only in preview

 1026 - TypeChanged
    The new version has a different type 'object' than the previous one ''.
    New: Azure.Search/preview/2024-05-01-preview/searchindex.json#L1107:13
    Old: Azure.Search/stable/2023-11-01/searchindex.json#L1005:13
40 occurrences.
==>(Will fix in next GA) Added PBI to use error definition from common type like in preview version 

 1038 - AddedPath, 1041 - AddedPropertyInResponse
 ==> (Ignore) New paths/Properties in preview

1032 - DifferentAllOf, 1020 - AddedEnumValue, 1041 - AddedPropertyInResponse,  1045 - AddedOptionalProperty
==> Added new enum in this preview

sazeesha098 commented:
> I understand that but why is the service team introducing these breaks? We don't want to break customers. So breaks must all be justified before they can be approved.

@JeffreyRichter After addressing some of the identified issues, here is a plan to fix furthers errors when comparing Stable to Preview. 

1034 - AddedRequiredProperty and 1025 - RequiredStatusChange: We've initiated a Product Backlog Item (PBI) to ensure skillsetCount is required in the next stable release.

1026 - TypeChanged, 1017 - ReferenceRedirection, and 1033 - RemovedProperty: We've also added a PBI to use error definitions from common types, aligning with newer preview versions.

Additionally, the error 1042 - ChangedParameterOrder shouldn't raise alerts in subsequent previews release. While we've aligned parameter ordering between this and previous stable version.

Furthermore, the following warnings can be disregarded as they pertain to additions in the preview version:
1038 - AddedPath
1032 - DifferentAllOf
1020 - AddedEnumValue
1041 - AddedPropertyInResponse
1045 - AddedOptionalProperty

sazeesha committed:
supporting the new binary packed type for binary embedding models

sazeesha committed:
Adding Edm.BitPackedUint8 to SearchFieldDataType

JeffreyRichter commented:
Please come to a Breaking Change Office Hours meeting on Mondays @ 10am pacific to discuss. If you come on April 1st, I will not be there but someone else can help you.

sazeesha committed:
Revert "Adding Edm.BitPackedUint8 to SearchFieldDataType"

This reverts commit 52711aa06dfb26ccf55669cd45f062b78ff645c8.

sazeesha committed:
Revert "supporting the new binary packed type for binary embedding models"

This reverts commit c28bf11ec14ee3ecb12cdf6b22a5b6527be7191d.

sazeesha098 commented:
> > I understand that but why is the service team introducing these breaks? We don't want to break customers. So breaks must all be justified before they can be approved.
> 
> @JeffreyRichter After addressing some of the identified issues, here is a plan to fix furthers errors when comparing Stable to Preview.
> 
> 1034 - AddedRequiredProperty and 1025 - RequiredStatusChange: We've initiated a Product Backlog Item (PBI) to ensure skillsetCount is required in the next stable release.
> 
> 1026 - TypeChanged, 1017 - ReferenceRedirection, and 1033 - RemovedProperty: We've also added a PBI to use error definitions from common types, aligning with newer preview versions.
> 
> Additionally, the error 1042 - ChangedParameterOrder shouldn't raise alerts in subsequent previews release. While we've aligned parameter ordering between this and previous stable version.
> 
> Furthermore, the following warnings can be disregarded as they pertain to additions in the preview version: 1038 - AddedPath 1032 - DifferentAllOf 1020 - AddedEnumValue 1041 - AddedPropertyInResponse 1045 - AddedOptionalProperty

@mikekistler thanks for helping us in the breaking changes office hours.

Here is the PR where replacing search error was approved in the previous preview version: https://github.com/Azure/azure-rest-api-specs/pull/27235

![image](https://github.com/Azure/azure-rest-api-specs/assets/52990439/2111acf4-f997-4c9d-a966-dd941757ab11)

sazeesha098 committed:
Merge branch 'main' into sazeesha/search-2024-05-01-preview

Commit:
Add new binary packed type for binary embedding models

Commit:
update comment desc

robertklee commented on specification/search/data-plane/Azure.Search/preview/2024-05-01-preview/searchservice.json:
```diff
@@ -6337,27 +6361,36 @@
+          {
+            "value": "hamming",
+            "name": "Hamming",
+            "description": "Only applicable to binary-derived data types. Determines dissimilarity by counting differing positions in binary vectors. The fewer differences, the closer the similarity."
```
adjusted the message to be a bit more consistent :)

Commit:
Add modelName to AmlVectorizer

jonathanserbent committed:
Merge branch 'main' into sazeesha/search-2024-05-01-preview

jonathanserbent committed:
Merge branch 'main' into sazeesha/search-2024-05-01-preview

Commit:
running prettier

Commit:
adding back cohere

mikekistler commented:
Adding the signedOff label based on approvals communicated over email.

jonathanserbent committed:
Merge branch 'main' into sazeesha/search-2024-05-01-preview

jonathanserbent committed:
Merge branch 'main' into sazeesha/search-2024-05-01-preview

jonathanserbent committed:
Merge branch 'main' into sazeesha/search-2024-05-01-preview

jonathanserbent committed:
Merge branch 'main' into sazeesha/search-2024-05-01-preview

sazeesha098 committed:
Merge branch 'main' into sazeesha/search-2024-05-01-preview
